### PR TITLE
Modified scripts for better fuzzy matching

### PR DESCRIPTION
**Pull request by pushpak-gote-mapup**
### **User description**
Added scripts for matching fuzzy similarity scores from new OSM columns and state neighboring roads.
Added feature: Create bridge statistics

___

### **PR Type**
enhancement, bug fix

___

### **Description**
- Added new functions to calculate similarity scores between multiple columns and a fixed column.
- Introduced functionality to read specific columns from a large CSV file to optimize memory usage.
- Enhanced error handling and logging across the scripts.
- Refactored the main processing logic to integrate new similarity calculations and data merging.
- Improved data filtering and output generation logic to ensure accurate results.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>calculate_match_percentage.py</strong><dd><code>Enhanced similarity calculation and CSV reading functions</code></dd></summary>
<hr>

hydrography-approach/processing_scripts/associate_data/calculate_match_percentage.py

<li>Added new function <code>calculate_similarity</code> to compute similarity scores.<br> <li> Introduced <code>read_exploded_osm_data_csv</code> to read specific columns from a <br>large CSV file.<br> <li> Modified <code>run</code> function to include new similarity calculations and data <br>merging.<br> <li> Enhanced error handling and logging.<br>

</details>

  </td>
  <td><a href="https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/34/files#diff-17e08b83eaff05ff944cab3bb74365b58c2737b1767fbdd321c71df9ed216515">+77/-18</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>get_merged_association_output.py</strong><dd><code>Refactored and enhanced similarity calculations and data merging</code></dd></summary>
<hr>

merge-approaches/get_merged_association_output.py

<li>Refactored similarity calculation functions for better error handling.<br> <li> Added <code>calculate_similarity_for_neighbouring_roads</code> to handle <br>neighbouring roads similarity.<br> <li> Updated <code>main</code> function to integrate new similarity calculations and <br>data merging.<br> <li> Improved data filtering and output generation logic.<br>

</details>

  </td>
  <td><a href="https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/34/files#diff-aa70096612983cc349ed7aa3c47d69b22b673c4ec77c722ea7d6e7c77051d945">+153/-231</a></td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by pushpak-gote-mapup** [comment]
Addressing PR agent's reviews.

**Comment by pushpak-gote-mapup** [comment]
[Error Handling](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/34/files#diff-17e08b83eaff05ff944cab3bb74365b58c2737b1767fbdd321c71df9ed216515R28-R30)
The error handling in the calculate_similarity and read_exploded_osm_data_csv functions uses a generic Exception catch, which might obscure the underlying issue. It's recommended to catch more specific exceptions or to at least log the stack trace along with the error message.
**Solved**
Added specific exceptions:
```python
def read_exploded_osm_data_csv(exploded_osm_data_csv: str, osm_cols_for_road_names: List[str]) -> Tuple[pd.DataFrame, List[str]]:
    """
    Reads the 'exploded_osm_data' CSV file and returns a DataFrame with the required columns.

    Parameters:
        exploded_osm_data_csv (str): The path to the CSV file.
        osm_cols_for_road_names (list): The list of column names to read from the CSV file.

    Returns:
        pandas.DataFrame: The DataFrame containing the required columns.
    """
    series_list=[]
    available_osm_road_names=[]
    for col in osm_cols_for_road_names:
        try:
            series=pd.read_csv(exploded_osm_data_csv,usecols=[col])
            series_list.append(series)
            available_osm_road_names.append(col)
        except ValueError as e:
            logger.warning(f"Column {col} not found in CSV: {e}")
            raise
        except pd.errors.EmptyDataError as e:
            logger.error(f"Empty data error when reading CSV: {e}", exc_info=True)
            raise
        except Exception as e:
            logger.error(f"Unexpected error when reading CSV: {e}", exc_info=True)
            raise
    
    
    try:
        exploded_osm_data_df=pd.concat(series_list,axis=1)
        return exploded_osm_data_df, available_osm_road_names
    except ValueError as e:
        logger.error(f"ValueError when concatenating series: {e}", exc_info=True)
        raise
    except Exception as e:
        logger.error(f"Unexpected error when concatenating series: {e}", exc_info=True)
        raise
    
``` 

[Performance Concern](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/34/files#diff-17e08b83eaff05ff944cab3bb74365b58c2737b1767fbdd321c71df9ed216515R21-R27)
The function calculate_similarity iterates over columns and performs fuzzy matching within a loop, which can be inefficient with large datasets. Consider vectorizing this operation or using more efficient data structures like numpy arrays or pandas methods that are optimized for performance.
**Solved**
Now using vectorized implementation:
```python
def vectorized_fuzz(s1: pd.Series, s2: pd.Series) -> pd.Series:
        try:
            mask = s1.notna() & s2.notna()
            result = pd.Series(0, index=s1.index)
            if mask.any():
                result[mask] = np.vectorize(fuzz.token_sort_ratio, otypes=[np.int64])(
                    s1[mask].astype(str), s2[mask].astype(str)
                )
            return result
        except ValueError as e:
            logger.error(f"ValueError in vectorized_fuzz: {str(e)}", exc_info=True)
            raise 
        except TypeError as e:
            logger.error(f"TypeError in vectorized_fuzz: {str(e)}", exc_info=True)
            raise 
        except Exception as e:
            logger.error(f"Unexpected error in vectorized_fuzz: {str(e)}", exc_info=True)
            raise 
``` 

[Exception Handling](https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/34/files#diff-aa70096612983cc349ed7aa3c47d69b22b673c4ec77c722ea7d6e7c77051d945R19-R23)
Similar to the first file, the use of generic Exception handling in functions like calculate_osm_similarity and read_geopackage_to_dataframe can be improved by specifying the type of exceptions expected or by enhancing the logging mechanism to provide more context about the errors.
**Solved**
Now using specific exceptions:
```python
def read_geopackage_to_dataframe(filepath: str) -> gpd.GeoDataFrame:
    """
    Read a GeoPackage file into a GeoDataFrame.

    Args:
        filepath (str): Path to the GeoPackage file.

    Returns:
        gpd.GeoDataFrame: The read GeoDataFrame.
    """
    try:
        return gpd.read_file(filepath)
    except FileNotFoundError as e:
        logger.error(f"FileNotFoundError: GeoPackage file not found at {filepath}: {str(e)}", exc_info=True)
        raise
    except PermissionError as e:
        logger.error(f"PermissionError: Unable to access GeoPackage file at {filepath}: {str(e)}", exc_info=True)
        raise
    except gpd.io.file.DriverError as e:
        logger.error(f"DriverError: Unable to read GeoPackage file at {filepath}: {str(e)}", exc_info=True)
        raise
    except Exception as e:
        logger.error(f"Unexpected error while reading GeoPackage file at {filepath}: {str(e)}", exc_info=True)
        raise
``` 

In addition to this, PR code suggestions are also implemented.

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        logger.error(f"{gpkg_file} is not a valid GeoPackage file: {str(e)}", exc_info=True)
+        raise ValueError(f"{gpkg_file} is not a valid GeoPackage file.")
+    except Exception as e:
+        logger.error(f"Unexpected error while reading GeoPackage file {gpkg_file}: {str(e)}", exc_info=True)
+        raise
+    
+
+def get_csv_length(csv_file: str) -> int:
+    """
+    Reads a GeoPackage file and returns its length.
```
Comment (varun-andhra-mapup) [inline hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py]: The docstring for get_csv_length incorrectly states it reads a GeoPackage file. It should specify it reads a CSV file.

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        raise FileNotFoundError(f"File {csv_file} does not exist.")
+    except ValueError as e:
+        logger.error(f"{csv_file} is not a valid CSV file: {str(e)}", exc_info=True)
+        raise ValueError(f"{csv_file} is not a valid CSV file.")
+    except Exception as e:
+        logger.error(f"Unexpected error while reading CSV file {csv_file}: {str(e)}", exc_info=True)
+        raise
+    
+
+def update_stats(stats: pd.DataFrame, description: str, count: int, stats_list: List[int]) -> Tuple[pd.DataFrame, List[int]]:
```
Comment (varun-andhra-mapup) [inline hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py:67]: The KeyError is logged and raised twice in a similar manner. Consider consolidating this into a single catch block.

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        return stats, stats_list
+    except KeyError as e:
+        logger.error(f"Description '{description}' does not exist in the DataFrame. {str(e)}", exc_info=True)
+        raise KeyError(f"Description '{description}' does not exist in the DataFrame.") from e
+    except Exception as e:
+        logger.error(f"Unexpected error in update_stats: {str(e)}", exc_info=True)
+        raise
+
+
+def create_bridge_statistics(
```
Comment (varun-andhra-mapup) [inline hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py:117]: The pattern for calculating bridge statistics is repetitive. Consider refactoring it to reduce redundancy. You could encapsulate the logic for calculating and updating statistics in a helper function.

**Review by varun-andhra-mapup — CHANGES_REQUESTED** [review]

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py (hunk 1 above)
Comment (pushpak-gote-mapup) [reply hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py]: Corrected it.

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py (hunk 2 above)
Comment (pushpak-gote-mapup) [reply hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py:67]: Only one KeyError is used now

File: hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py (hunk 3 above)
Comment (pushpak-gote-mapup) [reply hydrography-approach/processing_scripts/bridge_statistics/create_bridge_stats.py:117]: Done

**Review by varun-andhra-mapup — APPROVED** [review]

**Pull request merged**